From 03e34171efa61382ce6fda7b996ac75610e08949 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 24 Nov 2024 02:07:40 +0100
Subject: [PATCH] [Dogecash] Use Dogecoin's default fees (#83)

Dogecoin fees are significantly higher than on XEC.

These are the default settings:
- `relaymintxfee`: 0.001 DOGE/kB
- `blockmintxfee`: 0.001 DOGE/kB
- `paytxfee`: 0.01 DOGE/kB (which still seems to be below what's needed to get it mined)

diff --git a/src/policy/policy.h b/src/policy/policy.h
index 0397171f90afa8edde824a8fa1df1528eeb8e15c..87bda73a407aa3f792d8ea3f60dfa51c27fdca2e 100644
--- a/src/policy/policy.h
+++ b/src/policy/policy.h
@@ -27,7 +27,7 @@ static constexpr uint64_t DEFAULT_MAX_GENERATED_BLOCK_SIZE{ONE_MEGABYTE - 1000};
  * Default for -blockmintxfee, which sets the minimum feerate for a transaction
  * in blocks created by mining code.
  */
-static constexpr Amount DEFAULT_BLOCK_MIN_TX_FEE_PER_KB(1000 * SATOSHI);
+static constexpr Amount DEFAULT_BLOCK_MIN_TX_FEE_PER_KB(COIN / 1000);
 /**
  * The maximum size for transactions we're willing to relay/mine.
  */
@@ -64,7 +64,7 @@ static constexpr bool DEFAULT_PERMIT_BAREMULTISIG{true};
 static constexpr Amount DUST_RELAY_TX_FEE(3000 * SATOSHI);
 
 /** Default for -minrelaytxfee, minimum relay fee for transactions */
-static constexpr Amount DEFAULT_MIN_RELAY_TX_FEE_PER_KB(1000 * SATOSHI);
+static constexpr Amount DEFAULT_MIN_RELAY_TX_FEE_PER_KB(COIN / 1000);
 
 /**
  * When transactions fail script evaluations under standard flags, this flagset
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index 85acc174831ae62a7f0f2bc5749e40f25f8b7de7..79f43c7889096900a64259a5f9922613fc01bb28 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -95,7 +95,7 @@ MakeWalletDatabase(const std::string &name, const DatabaseOptions &options,
                    DatabaseStatus &status, bilingual_str &error);
 
 //! -paytxfee default
-constexpr Amount DEFAULT_PAY_TX_FEE = Amount::zero();
+constexpr Amount DEFAULT_PAY_TX_FEE = COIN / 100;
 //! -fallbackfee default
 static const Amount DEFAULT_FALLBACK_FEE = Amount::zero();
 //! -mintxfee default
@@ -120,9 +120,9 @@ static const bool DEFAULT_SPEND_ZEROCONF_CHANGE = true;
 static const bool DEFAULT_WALLETBROADCAST = true;
 static const bool DEFAULT_DISABLE_WALLET = false;
 //! -maxtxfee default
-constexpr Amount DEFAULT_TRANSACTION_MAXFEE{COIN / 10};
+constexpr Amount DEFAULT_TRANSACTION_MAXFEE{100 * COIN};
 //! Discourage users to set fees higher than this amount (in satoshis) per kB
-constexpr Amount HIGH_TX_FEE_PER_KB{COIN / 100};
+constexpr Amount HIGH_TX_FEE_PER_KB{10 * COIN};
 //! -maxtxfee will warn if called with a higher fee than this amount (in
 //! satoshis)
 constexpr Amount HIGH_MAX_TX_FEE{100 * HIGH_TX_FEE_PER_KB};
diff --git a/test/functional/dogecoin_default_fees.py b/test/functional/dogecoin_default_fees.py
new file mode 100644
index 0000000000000000000000000000000000000000..31d54d6ea0e0e5ebf01e851397c8d71f636b6a89
--- /dev/null
+++ b/test/functional/dogecoin_default_fees.py
@@ -0,0 +1,51 @@
+# Copyright (c) 2024 The Bitcoin developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test default dogecoin fees.
+"""
+
+from decimal import Decimal
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, remove_from_config
+
+
+class DogecoinFeesTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+        self.extra_args = [["-usedogeunit=1"]]
+        self.rpc_timeout = 240
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        node = self.nodes[0]
+
+        # Remove test framework configs added in util.py
+        remove_from_config(
+            node.datadir, ["minrelaytxfee=10", "blockmintxfee=10", "fallbackfee=200"]
+        )
+
+        self.restart_node(0)
+
+        assert_equal(node.getnetworkinfo()["relayfee"], Decimal("0.001"))
+        assert_equal(node.getmempoolinfo()["minrelaytxfee"], Decimal("0.001"))
+        assert_equal(node.getmempoolinfo()["mempoolminfee"], Decimal("0.001"))
+
+        address = node.getnewaddress()
+        address2 = node.getnewaddress()
+
+        self.generatetoaddress(node, 110, address)
+
+        txid = node.sendtoaddress(address2, 10)
+
+        # -paytxfee is 10x that of the relay fee
+        tx = node.gettransaction(txid)
+        assert_equal(tx["fee"], Decimal("-0.0022500"))
+
+
+if __name__ == "__main__":
+    DogecoinFeesTest().main()
diff --git a/test/functional/feature_notifications.py b/test/functional/feature_notifications.py
index 3bd9698daeae730a9cb07ad6d8f20fec4189a901..2825fd284db2b97a28285f7aa1700fcda2712d3b 100644
--- a/test/functional/feature_notifications.py
+++ b/test/functional/feature_notifications.py
@@ -56,6 +56,7 @@ class NotificationsTest(BitcoinTestFramework):
                 f"-alertnotify=echo > {os.path.join(self.alertnotify_dir, '%s')}",
                 f"-blocknotify=echo > {os.path.join(self.blocknotify_dir, '%s')}",
                 f"-shutdownnotify=echo > {self.shutdownnotify_file}",
+                "-paytxfee=10",
             ],
             [
                 "-rescan",
@@ -63,6 +64,7 @@ class NotificationsTest(BitcoinTestFramework):
                     "-walletnotify=echo %h_%b >"
                     f" {os.path.join(self.walletnotify_dir, notify_outputname('%w', '%s'))}"
                 ),
+                "-paytxfee=10",
             ],
         ]
         self.wallet_names = [self.default_wallet_name, self.wallet]
diff --git a/test/functional/interface_bitcoin_cli.py b/test/functional/interface_bitcoin_cli.py
index ab4112766f8248c38062ae43da477014d6304440..6b2ceb9d8012c006239a75655d7bcdf8377c641d 100644
--- a/test/functional/interface_bitcoin_cli.py
+++ b/test/functional/interface_bitcoin_cli.py
@@ -34,6 +34,7 @@ class TestBitcoinCli(BitcoinTestFramework):
     def set_test_params(self):
         self.setup_clean_chain = True
         self.num_nodes = 1
+        self.extra_args = [["-paytxfee=10"]]
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_cli()
diff --git a/test/functional/p2p_ibd_txrelay.py b/test/functional/p2p_ibd_txrelay.py
index a671544688214260c756e69e1e2fbb8365504693..bdfa377591ebac068b3ce7a704e447a8e22cd93b 100644
--- a/test/functional/p2p_ibd_txrelay.py
+++ b/test/functional/p2p_ibd_txrelay.py
@@ -27,7 +27,7 @@ from test_framework.p2p import (
 )
 from test_framework.test_framework import BitcoinTestFramework
 
-MAX_FEE_FILTER = Decimal(9170997) / XEC
+MAX_FEE_FILTER = Decimal(9452957) / XEC
 NORMAL_FEE_FILTER = Decimal(100) / XEC
 
 
diff --git a/test/functional/rpc_fundrawtransaction.py b/test/functional/rpc_fundrawtransaction.py
index 6ecb42f79a7acbedac1930ab6357821ed900008d..451ee7016dbda619f0097a87eb2b08b2d3e93d67 100644
--- a/test/functional/rpc_fundrawtransaction.py
+++ b/test/functional/rpc_fundrawtransaction.py
@@ -30,7 +30,9 @@ class RawTransactionsTest(BitcoinTestFramework):
         self.setup_clean_chain = True
         # This test isn't testing tx relay. Set whitelist on the peers for
         # instant tx relay.
-        self.extra_args = [["-whitelist=noban@127.0.0.1"]] * self.num_nodes
+        self.extra_args = [
+            ["-whitelist=noban@127.0.0.1", "-maxtxfee=100000"]
+        ] * self.num_nodes
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
diff --git a/test/functional/rpc_psbt.py b/test/functional/rpc_psbt.py
index 8ac113c5f803a37f796a1ac075c6920a00bf417c..6c58de110aa270e421d9c318d3cedba5b9827744 100644
--- a/test/functional/rpc_psbt.py
+++ b/test/functional/rpc_psbt.py
@@ -23,6 +23,7 @@ class PSBTTest(BitcoinTestFramework):
         self.num_nodes = 3
         self.supports_cli = False
         self.noban_tx_relay = True
+        self.extra_args = [["-maxtxfee=100000"]] * 3
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
diff --git a/test/functional/test_framework/util.py b/test/functional/test_framework/util.py
index 7dbd86009ee7e7e33466b01df91f6aeb73571de8..f01abd9a2c827197a2c16c7dcc8fe07cc067518f 100644
--- a/test/functional/test_framework/util.py
+++ b/test/functional/test_framework/util.py
@@ -579,6 +579,9 @@ def write_config(config_path, *, n, chain, extra_config="", disable_autoconnect=
         f.write("natpmp=0\n")
         f.write("usecashaddr=1\n")
         f.write("legacyscriptrules=0\n")
+        f.write("usedogeunit=0\n")
+        f.write("minrelaytxfee=10\n")
+        f.write("blockmintxfee=10\n")
         # Increase peertimeout to avoid disconnects while using mocktime.
         # peertimeout is measured in mock time, so setting it large enough to
         # cover any duration in mock time is sufficient. It can be overridden
@@ -620,6 +623,16 @@ def append_config(datadir, options):
             f.write(f"{option}\n")
 
 
+def remove_from_config(datadir, options):
+    with open(config_path(datadir), "r", encoding="utf8") as f:
+        config = f.read()
+        for option in options:
+            config = config.replace(option + "\n", "")
+
+    with open(config_path(datadir), "w", encoding="utf8") as f:
+        f.write(config)
+
+
 def get_auth_cookie(datadir, chain):
     user = None
     password = None
diff --git a/test/functional/wallet_avoidreuse.py b/test/functional/wallet_avoidreuse.py
index a9540da375a9a9e95a3a155d166a6b002eea0f30..0a24c7af87cd2fbbc00bff32f621ab60fcaf2cf7 100644
--- a/test/functional/wallet_avoidreuse.py
+++ b/test/functional/wallet_avoidreuse.py
@@ -78,6 +78,7 @@ class AvoidReuseTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
         self.noban_tx_relay = True
+        self.extra_args = [["-paytxfee=10"]] * 2
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
diff --git a/test/functional/wallet_basic.py b/test/functional/wallet_basic.py
index 7f2d540bc7c6c586ded03501c6f153aee9288c25..7c86d2e10e52120188012831f41f46a63c5fe8e3 100644
--- a/test/functional/wallet_basic.py
+++ b/test/functional/wallet_basic.py
@@ -24,7 +24,7 @@ class WalletTest(BitcoinTestFramework):
         self.noban_tx_relay = True
         self.setup_clean_chain = True
         self.extra_args = [
-            ["-acceptnonstdtxn=1"],
+            ["-acceptnonstdtxn=1", "-paytxfee=10"],
         ] * self.num_nodes
         self.supports_cli = False
 
diff --git a/test/functional/wallet_create_tx.py b/test/functional/wallet_create_tx.py
index 9afdf78c14505fec8f6df8f3cef885a37456e108..be6533ae1693ce2541768f334f6b8f5336b23e1d 100644
--- a/test/functional/wallet_create_tx.py
+++ b/test/functional/wallet_create_tx.py
@@ -4,7 +4,7 @@
 
 from test_framework.blocktools import TIME_GENESIS_BLOCK
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import assert_equal, assert_raises_rpc_error
+from test_framework.util import append_config, assert_equal, assert_raises_rpc_error
 
 
 class CreateTxWalletTest(BitcoinTestFramework):
@@ -21,6 +21,8 @@ class CreateTxWalletTest(BitcoinTestFramework):
         self.generate(self.nodes[0], 200)
         self.nodes[0].setmocktime(0)
 
+        append_config(self.nodes[0].datadir, ["maxtxfee=100000"])
+
         self.test_anti_fee_sniping_disabled()
         self.test_tx_size_too_large()
 
diff --git a/test/functional/wallet_groups.py b/test/functional/wallet_groups.py
index 223e96706081d807ecb8f52062588b386fadd8ed..04fb940a5e53a7ecb3b8a2d6a4617cd1b754a651 100644
--- a/test/functional/wallet_groups.py
+++ b/test/functional/wallet_groups.py
@@ -15,15 +15,15 @@ class WalletGroupTest(BitcoinTestFramework):
         self.num_nodes = 5
         self.noban_tx_relay = True
         self.extra_args = [
-            [],
-            [],
-            ["-avoidpartialspends"],
+            ["-paytxfee=10"],
+            ["-paytxfee=10"],
+            ["-avoidpartialspends", "-paytxfee=10"],
             # 2.93 XEC is the threshold that causes a node to prefer a
             # non-grouped tx (3 inputs, 2 outputs) to a grouped tx (5 inputs,
             # 2 outputs). The fee for the grouped tx is 294 sats higher than
             # the fee for the non-grouped tx. See tx5 below.
-            ["-maxapsfee=2.93"],
-            ["-maxapsfee=2.94"],
+            ["-maxapsfee=2.93", "-paytxfee=10"],
+            ["-maxapsfee=2.94", "-paytxfee=10"],
         ]
 
         self.rpc_timeout = 480
diff --git a/test/functional/wallet_importdescriptors.py b/test/functional/wallet_importdescriptors.py
index 1d70e786e81c2e198c0be2d8f007a41b3df45417..b2352cad14294037162a0e3b320cc635e8b57960 100644
--- a/test/functional/wallet_importdescriptors.py
+++ b/test/functional/wallet_importdescriptors.py
@@ -30,7 +30,7 @@ class ImportDescriptorsTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
         self.noban_tx_relay = True
-        self.extra_args = [[], ["-keypool=5"]]
+        self.extra_args = [["-paytxfee=10"], ["-keypool=5", "-paytxfee=10"]]
         self.setup_clean_chain = True
 
     def skip_test_if_missing_module(self):
diff --git a/test/functional/wallet_multiwallet.py b/test/functional/wallet_multiwallet.py
index 3732a15415dd4e41f385548d5afb5f0b433c9fd9..1e166bb34509f0c6b4ce82187729639adedd94e9 100644
--- a/test/functional/wallet_multiwallet.py
+++ b/test/functional/wallet_multiwallet.py
@@ -364,11 +364,13 @@ class MultiWalletTest(BitcoinTestFramework):
         assert_equal(batch[0]["result"]["chain"], self.chain)
         assert_equal(batch[1]["result"]["walletname"], "w1")
 
-        self.log.info("Check for per-wallet settxfee call")
-        assert_equal(w1.getwalletinfo()["paytxfee"], 0)
-        assert_equal(w2.getwalletinfo()["paytxfee"], 0)
+        self.log.info(
+            "Check for per-wallet settxfee call (paytxfee defaults to 10000 on Dogecoin)"
+        )
+        assert_equal(w1.getwalletinfo()["paytxfee"], 10000)
+        assert_equal(w2.getwalletinfo()["paytxfee"], 10000)
         w2.settxfee(1000)
-        assert_equal(w1.getwalletinfo()["paytxfee"], 0)
+        assert_equal(w1.getwalletinfo()["paytxfee"], 10000)
         assert_equal(w2.getwalletinfo()["paytxfee"], Decimal("1000.00"))
 
         self.log.info("Test dynamic wallet loading")
